File: kernel/type.py
# Author: Bohua Zhan

from collections import UserDict

from logic import term_ord
from util import typecheck


class TypeException(Exception):
    """Indicates error in processing types."""
    def __init__(self, msg):
        self.msg = msg

    def __str__(self):
        return self.msg


class TypeMatchException(Exception):
    """Indicates error when matching types."""
    def __init__(self, msg):
        self.msg = msg

    def __str__(self):
        return self.msg


class TyInst(UserDict):
    """Instantiation of schematic type variables."""
    def __str__(self):
        return ', '.join("'%s := %s" % (nm, T) for nm, T in self.items())

    def __copy__(self):
        return TyInst(self)


"""Default parser for types. If None, Type() is unable to parse type."""
type_parser = None

"""Default printer for types. If None, Type.print_basic is used."""
type_printer = None


class Type():
    """Represents a type in higher-order logic.
    
    Types in HOL are formed by two kinds of constructors: STVar, TVar and TConst.

    STVar(name) represents a schematic type variable with the given name.
    TVar(name) represents a type variable with the given name. TConst(f, args)
    represents a type constant applied to a list of arguments.
    
    There are two fundamental type constants:
    
    - booleans, with name "bool" and no arguments.
    
    - functions, with name "fun" and two arguments: the domain and codomain
    types. TConst("fun", a, b) is printed as a => b. The => sign associates to
    the right.
    
    Further defined type constants include:
    
    - natural numbers, with name "nat" and no arguments.
    
    - lists, with name "list" and one argument.

    - product, with name "prod" and two arguments. TConst("prod", a, b) is
    printed as a * b.
    
    Examples:
    
    nat => bool: functions from natural numbers to booleans (predicates on
    natural numbers).

    nat => nat: functions from natural numbers to natural numbers.

    nat => nat => nat: or nat => (nat => nat), functions from two natural
    numbers to natural numbers.

    nat * nat => nat: functions from a pair of natural numbers to natural
    numbers.

    nat list: list of natural numbers.

    nat list list: list of lists of natural numbers.

    """
    # ty values for distinguishing between Type objects.
    STVAR, TVAR, TCONST = range(3)

    def __init__(self, arg):
        if not isinstance(arg, Type):
            if type_parser is not None:
                T = type_parser(arg)
            else:
                raise TypeException('Type: parser not found.')
        else:
            T = arg

        # Now copy the content of T onto self
        self.__dict__.update(T.__dict__)

    def is_stvar(self):
        """Return whether self is a schematic type variable."""
        return self.ty == Type.STVAR

    def is_tvar(self):
        """Return whether self is a type variable."""
        return self.ty == Type.TVAR

    def is_tconst(self):
        """Return whether self is given by a type constructor."""
        return self.ty == Type.TCONST

    def is_fun(self):
        """Whether self is of the form a => b."""
        return self.is_tconst() and self.name == "fun"
    
    def domain_type(self):
        """Given a type of form a => b, return a."""
        assert self.is_fun(), "domain_type"
        return self.args[0]
    
    def range_type(self):
        """Given a type of form a => b, return b."""
        assert self.is_fun(), "range_type"
        return self.args[1]

    def strip_type(self):
        """Given a type of form a_1 => ... => a_n, b, return the pair
        [a_1, ... a_n], b.
        
        """
        if self.is_fun():
            domains, range = self.range_type().strip_type()
            return ([self.domain_type()] + domains, range)
        else:
            return ([], self)
        
    def print_basic(self):
        """Basic printing function for types."""
        if self.is_stvar():
            return "?'" + self.name
        elif self.is_tvar():
            return "'" + self.name
        elif self.is_tconst():
            if len(self.args) == 0:
                return self.name
            elif len(self.args) == 1:
                # Insert parenthesis if the single argument is a function.
                if self.args[0].is_fun():
                    return "(%s) %s" % (self.args[0], self.name)
                else:
                    return "%s %s" % (self.args[0], self.name)
            elif self.is_fun():
                # 'a => 'b => 'c associates to the right. So parenthesis is
                # needed to express ('a => 'b) => 'c.
                if self.args[0].is_fun():
                    return "(%s) => %s" % (self.args[0], self.args[1])
                else:
                    return "%s => %s" % (self.args[0], self.args[1])
            else:
                return "(%s) %s" % (", ".join(str(t) for t in self.args), self.name)
        else:
            raise TypeError

    def __str__(self):
        if type_printer is None:
            return self.print_basic()
        else:
            return type_printer(self)

    def __repr__(self):
        if self.is_stvar():
            return "STVar(%s)" % self.name
        if self.is_tvar():
            return "TVar(%s)" % self.name
        elif self.is_tconst():
            return "TConst(%s, %s)" % (self.name, list(self.args))
        else:
            raise TypeError

    def __hash__(self):
        if not hasattr(self, "_hash_val"):
            if self.is_stvar():
                self._hash_val = hash(("STVAR", self.name))
            if self.is_tvar():
                self._hash_val = hash(("TVAR", self.name))
            elif self.is_tconst():
                self._hash_val = hash(("TCONST", self.name, tuple(hash(arg) for arg in self.args)))
        return self._hash_val
    
    def __eq__(self, other):
        if other is None:
            return False
        assert isinstance(other, Type), "cannot compare Type with %s" % str(type(other))

        if id(self) == id(other):
            return True

        if self.ty != other.ty:
            return False
        elif self.is_stvar() or self.is_tvar():
            return self.name == other.name
        elif self.is_tconst():
            return self.name == other.name and self.args == other.args
        else:
            raise TypeError

    def __le__(self, other):
        """Fast version of comparison."""
        if self.ty != other.ty:
            return self.ty <= other.ty
        elif self.is_stvar() or self.is_tvar():
            return self.name <= other.name
        elif self.is_tconst():
            return (self.name, self.args) <= (other.name, other.args)
        else:
            raise TypeError

    def __lt__(self, other):
        """Fast version of comparison."""
        if self.ty != other.ty:
            return self.ty < other.ty
        elif self.is_stvar() or self.is_tvar():
            return self.name < other.name
        elif self.is_tconst():
            return (self.name, self.args) < (other.name, other.args)
        else:
            raise TypeError

    def size(self):
        """Return the size of the type."""
        if self.is_stvar() or self.is_tvar():
            return 1
        elif self.is_tconst():
            return 1 + sum(T.size() for T in self.args)
        else:
            raise TypeError

    def subst(self, tyinst=None, **kwargs):
        """Simultaneously substitute for the type variables using tyinst.
        
        Parameters
        ==========
        tyinst : TyInst
            Type instantiation to be substituted.

        """
        if tyinst is None:
            tyinst = TyInst(**kwargs)
        if self.is_stvar():
            if self.name in tyinst:
                return tyinst[self.name]
            else:
                return self
        elif self.is_tvar():
            return self
        elif self.is_tconst():
            return TConst(self.name, *(T.subst(tyinst) for T in self.args))
        else:
            raise TypeError

    def match_incr(self, T, tyinst):
        """Incremental type matching of self with T.
        
        Parameters
        ==========
        tyinst : TyInst
            The current instantiation. This is updated by the function.

        """
        if self.is_stvar():
            if self.name in tyinst:
                if T != tyinst[self.name]:
                    raise TypeMatchException('Unable to match %s with %s' % (T, tyinst[self.name]))
            else:
                tyinst[self.name] = T
        elif self.is_tvar():
            if self != T:
                raise TypeMatchException('Unable to match %s with %s' % (self, T))
        elif self.is_tconst():
            if (not T.is_tconst()) or T.name != self.name:
                raise TypeMatchException('Unable to match %s with %s' % (self, T))
            else:
                for arg, argT in zip(self.args, T.args):
                    arg.match_incr(argT, tyinst)
        else:
            raise TypeError

    def match(self, T):
        """Type matching of self with T.
        
        Return the resulting instantiation, or raise TypeMatchException
        if matching fails.

        """
        tyinst = TyInst()
        self.match_incr(T, tyinst)
        return tyinst

    def get_stvars(self):
        """Return the list of schematic type variables."""
        def collect(T):
            if T.is_stvar():
                return [T]
            elif T.is_tvar():
                return []
            else:
                return sum([collect(arg) for arg in T.args], [])

        return term_ord.sorted_typs(collect(self))

    def get_tvars(self):
        """Return the list of type variables."""
        def collect(T):
            if T.is_stvar():
                return []
            elif T.is_tvar():
                return [T]
            else:
                return sum([collect(arg) for arg in T.args], [])

        return term_ord.sorted_typs(collect(self))

    def get_tsubs(self):
        """Return the list of types appearing in self."""
        def collect(T):
            if T.is_stvar() or T.is_tvar():
                return [T]
            else:
                return sum([collect(arg) for arg in T.args], [T])

        return term_ord.sorted_typs(collect(self))

    def convert_stvar(self):
        if self.is_stvar():
            raise TypeException("convert_stvar")
        elif self.is_tvar():
            return STVar(self.name)
        elif self.is_tconst():
            return TConst(self.name, *(arg.convert_stvar() for arg in self.args))
        else:
            raise TypeError

    def is_numeral_type(self):
        return self in (NatType, IntType, RealType)


class STVar(Type):
    """Schematic type variable."""
    def __init__(self, name):
        self.ty = Type.STVAR
        self.name = name

class TVar(Type):
    """Type variable."""
    def __init__(self, name):
        self.ty = Type.TVAR
        self.name = name

class TConst(Type):
    """Type constant, applied to a list of arguments."""
    def __init__(self, name, *args):
        self.ty = Type.TCONST
        self.name = name
        self.args = args

def TFun(*args):
    """Returns the function type arg1 => arg2 => ... => argn."""
    typecheck.checkinstance('TFun', args, [Type])
    res = args[-1]
    for arg in reversed(args[:-1]):
        res = TConst("fun", arg, res)
    return res


# Boolean type
BoolType = TConst("bool")

# Numeral types
NatType = TConst('nat')
IntType = TConst('int')
RealType = TConst('real')

File: holsmt/type.cpp
#include<iostream>
#include<numeric>
#include<algorithm>
#include "type.h"

TConst BoolType = TConst("bool");
TConst NatType = TConst("nat");
TConst IntType = TConst("int");
TConst RealType = TConst("real");

std::ostream& operator << (std::ostream& os, const Type& t) {
	if (t.type() == 0) {
		return os << "?'" << t.name();
	}
	else if (t.type() == 1) {
		return os << "'" << t.name();
	}
	else if (t.type() == 2) {
		std::vector<Type> args = t.getArgs();
		if (args.size() == 0) {
			return os << t.name();
		}
		else if (args.size() == 1) {
			if (args[0].is_fun()) {
				return os << "(" << args[0] << ") " << t.name();
			}
			else {
				return os << args[0] << " " << t.name();
			}
		}
		else if (t.is_fun()) {
			if (args[0].is_fun()) {
				return os << "(" << args[0] << ") => " << args[1];
			}
			else {
				return  os << args[0] << " => " << args[1];
			}
		}
		else {
			os << "(";
			for (auto s : args) {
				if (s == args.back()) {
					os << s <<") ";
				}
				else {
					os << s << ", ";
				}
			}
			return os << t.name();
		}
	}
}

inline std::string Type::name() const {
	return na;
}

inline std::vector<Type> Type::getArgs() const {
	if (ty == 2) {
		return args;
	}
	else {
		throw std::runtime_error("Type error.");
	}
}

inline int Type::type() const {
	return ty;
}

inline bool Type::is_stvar() const {
	return ty == 0;
}

inline bool Type::is_tvar() const {
	return ty == 1;
}

inline bool Type::is_tconst() const {
	return ty == 2;
}

inline bool Type::is_fun() const {
	return ty == 2 && na == "fun";
}

inline Type Type::domain_type() const{
	assert(this->is_fun());
	return args[0];
}

inline Type Type::range_type() const{
	assert(this->is_fun());
	return args[1];
}

std::pair<std::vector<Type>, Type> Type::strip_type() const{
	if(this->is_fun()){
		std::pair<std::vector<Type>, Type> p = this->range_type().strip_type();
		
		std::vector<Type> domain_ty = {this->domain_type()};

		domain_ty.insert(domain_ty.end(), p.first.begin(), p.first.end());

		std::pair<std::vector<Type>, Type> v_new = {domain_ty, p.second};

		return v_new;
	}else{
		std::vector<Type> v;
		std::pair<std::vector<Type>, Type> p = {v, *this};
		return p;
	}
}

int Type::size() const {
	if(this->is_stvar() || this->is_tvar()){
		return 1;
	}
	else if(this->is_tconst()){
		int n = 1;
		for(auto a : args){
			n += a.size();
		}
		return n;
	}
	else{
		throw std::runtime_error("Type error");
	}
}

Type subst(Type& t, std::map<STVar, Type>& s) {
	if(t.is_stvar()){
		STVar st = to_stvar(t);
		if(s.find(st) != s.end()){
			return s[st];
		}else{
			return t;
		}
	}else if(t.is_tvar()){
		return t;
	}else if(t.is_tconst()){
		std::vector<Type> v_args;
		for(auto a : t.getArgs()){
			Type ta = subst(a, s);
			v_args.push_back(ta);
		}
		return TConst(t.name(), v_args);
	}
}

bool Type::operator==(const Type& t) const{
	if (this == &t) {
		return true;
	}
	if(ty != t.type())	{
		return false;
	}
	else if (ty == 0 || ty == 1) {
		return na == t.name();
	}
	else if (ty == 2) {
		return na == t.name() && args == t.getArgs();
	}
	else {
		throw std::runtime_error("Bad type.");
	}
}

bool Type::operator<(const Type& t) const{
	if(ty != t.type()){
		return ty < t.type();
	}else if(ty == 0 || ty == 1){
		return na < t.name();
	}else if(ty == 2){
		if(na != t.name()){
			return na < t.name();
		}
		else
		{
			int size_i = this->size();
			int size_j = t.size();
			std::vector<Type> args_i = this->getArgs();
			std::vector<Type> args_j = t.getArgs();
			for (int i = 0; i < std::min(size_i, size_j); ++i) {
				if (args_i[i] != args_j[i]) {
					return args_i[i] < args_j[i];
				}
			}
			return size_i < size_j;  
		}
	}
}

bool Type::operator!=(const Type& t) const {
	return !(*this == t);
}

inline STVar to_stvar(Type& t){
	assert(t.type() == 0);
	return STVar(t.name());
}

inline TVar to_tvar(Type& t){
	assert(t.type() == 1);
	return TVar(t.name());
}

inline TConst to_tconst(Type& t){
	assert(t.type() == 2);
	return TConst(t.name(), t.getArgs());
}

std::map<std::string, Type> Type::match_incr
	(Type& t, std::map<std::string, Type> m) const {
	if (this->is_stvar()) {
		if (m.find(this->name()) != m.end()) {
			if (t != m[this->name()]) {
				throw std::runtime_error("Match error");
			}
		}
		else {
			m.insert({ this->name(), t });
			return m;
		}
	}
	else if (this->is_tvar()) {
		if (*this != t) {
			throw std::runtime_error("Match error");
			return m;
		}
	}
	else if (this->is_tconst()) {
		if (!t.is_tconst() || t.name() != na) {
			throw std::runtime_error("Match error");
		}
		else {
			for (int i = 0; i < args.size(); ++i) {
				m =args[i].match_incr(t.getArgs()[i], m);
			}
			return m;
		}
	}
}

std::map<std::string, Type> Type::match(Type& t) const {
	std::map<std::string, Type> m;
	m = match_incr(t, m);
	return m;
}

std::set<Type> Type::get_stvars() const {
	std::set<Type> s;
	if (this->is_stvar()) {
		s.insert(*this);
		return s;
	}
	else if (this->is_tvar()) {
		return s;
	}
	else {
		for (auto a : this->getArgs()) {
			std::set<Type> s_arg = a.get_stvars();
			s.insert(s_arg.begin(), s_arg.end());
		}
		return s;
	}
}

std::set<Type> Type::get_tvars() const {
	std::set<Type> s;
	if (this->is_tvar()) {
		s.insert(*this);
		return s;
	}
	else if (this->is_stvar()) {
		return s;
	}
	else {
		for (auto a : this->getArgs()) {
			std::set<Type> s_arg = a.get_stvars();
			s.insert(s_arg.begin(), s_arg.end());
		}
		return s;
	}
}

std::set<Type> Type::get_tsubs() const {
	std::set<Type> s;
	if (this->is_tvar() || this->is_stvar()) {
		s.insert(*this);
		return s;
	}
	else {
		for (auto a : this->getArgs()) {
			std::set<Type> s_arg = a.get_stvars();
			s.insert(s_arg.begin(), s_arg.end());
		}
		return s;
	}
}

Type Type::convert_stvar() const {
	if (this->is_stvar()) {
		throw std::runtime_error("convert_stvar");
	}
	else if (this->is_tvar()) {
		return Type(0, na);
	}
	else if(this->is_tconst()) {
		std::vector<Type> v = this->getArgs();
		std::vector<Type> v_new;
		for (auto& a : v) {
			v_new.push_back(a.convert_stvar());
		}
		return TConst(na, v_new);
	}
}

bool Type::is_numeral_type() const {
	return *this == NatType || *this == IntType || *this == RealType;
}

Type TFun(std::vector<Type> v) {
	Type res = v.back();
	for (auto iter = v.crbegin() + 1; iter != v.crend(); ++iter) {
		res = TConst("fun", { *iter, res });
	}
	return res;
}